src/main/scala/mesosphere/marathon/core/storage/repository/impl/GroupRepositoryImpl.scala

```diff
@@ -0,0 +1,184 @@
+package mesosphere.marathon.core.storage.repository.impl
+
+// scalastyle:off
+import java.time.OffsetDateTime
+import java.time.format.DateTimeFormatter
+
+import akka.http.scaladsl.marshalling.Marshaller
+import akka.http.scaladsl.unmarshalling.Unmarshaller
+import akka.stream.Materializer
+import akka.stream.scaladsl.Source
+import akka.{ Done, NotUsed }
+import mesosphere.marathon.Protos
+import mesosphere.marathon.core.storage.repository.{ AppRepository, GroupRepository }
+import mesosphere.marathon.core.storage.store.impl.BasePersistenceStore
+import mesosphere.marathon.core.storage.store.impl.cache.{ LazyCachingPersistenceStore, LoadTimeCachingPersistenceStore }
+import mesosphere.marathon.core.storage.store.{ IdResolver, PersistenceStore }
+import mesosphere.marathon.state.{ AppDefinition, Group, PathId, Timestamp }
+
+import scala.async.Async.{ async, await }
+import scala.collection.JavaConverters._
+import scala.collection.immutable.Seq
+import scala.concurrent.{ ExecutionContext, Future }
+import scala.util.control.NonFatal
+// scalastyle:on
+
+private[storage] case class StoredGroup(
+    id: PathId,
+    appIds: Map[PathId, OffsetDateTime],
+    storedGroups: Seq[StoredGroup],
+    dependencies: Set[PathId],
+    version: OffsetDateTime) {
+  def resolve(
+    groupRepository: GroupRepository,
+    appRepository: AppRepository)(implicit ctx: ExecutionContext): Future[Group] = async {
+    val appFutures = appIds.map { case (appId, appVersion) => appRepository.getVersion(appId, appVersion) }
```

Comment:
Perhaps add a method `AppRepository.getVersions` so we can cap concurrent access?